### PR TITLE
[IOTDB-6142] Pipe: Implemented IoTDBOpcUaConnector to enable transfer using Opc Ua protocol

### PR DESCRIPTION
**Pull request by Caideyipi**
## IOTDB-6142
Enabled data sending to subscriber clients through Opc Ua protocol.

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/event/dml/insertion/TabletInsertionEvent.java [hunk 1]
```diff
@@ -44,4 +44,26 @@ public interface TabletInsertionEvent extends Event {
...
+   */
+  Tablet convertToTablet();
+
+  /**
+   * Parse the event to resolve its pattern. The parsing process is done in
+   * PipeRawTabletInsertionEvent.
+   *
+   * @return the converted PipeRawTabletInsertionEvent
+   */
+  TabletInsertionEvent parseEventWithPattern();
```
Comment (Caideyipi) [inline iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/event/dml/insertion/TabletInsertionEvent.java:61]: parseEventWithPatternOrTime

File: iotdb-core/datanode/pom.xml [hunk 2]
```diff
@@ -250,6 +250,11 @@
             <artifactId>openapi</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.milo</groupId>
+            <artifactId>sdk-server</artifactId>
+            <version>0.6.10</version>
```
Comment (Caideyipi) [inline iotdb-core/datanode/pom.xml:256]: "Version" should be defined in the profile scope

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/config/constant/PipeConnectorConstant.java [hunk 3]
```diff
@@ -54,6 +54,12 @@ public class PipeConnectorConstant {
   public static final String CONNECTOR_WEBSOCKET_PORT_KEY = "connector.websocket.port";
   public static final int CONNECTOR_WEBSOCKET_PORT_DEFAULT_VALUE = 8080;
 
+  public static final String CONNECTOR_IOTDB_OPC_UA_TCP_BIND_PORT_KEY = "connector.tcp.port";
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/config/constant/PipeConnectorConstant.java:57]: CONNECTOR_OPC_UA_TCP_BIND_PORT_KEY + "connector.opcua.tcp.port"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/config/constant/PipeConnectorConstant.java [hunk 4]
```diff
@@ -54,6 +54,12 @@ public class PipeConnectorConstant {
...
+  public static final int CONNECTOR_IOTDB_OPC_UA_TCP_BIND_PORT_DEFAULT_VALUE = 12686;
+
+  public static final String CONNECTOR_IOTDB_OPC_UA_HTTPS_BIND_PORT_KEY = "connector.https.port";
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/config/constant/PipeConnectorConstant.java:60]: CONNECTOR_OPC_UA_HTTPS_BIND_PORT_KEY + "connector.opcua.https.port"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/airgap/IoTDBAirGapConnector.java [hunk 5]
```diff
@@ -187,12 +187,7 @@ public void transfer(TabletInsertionEvent tabletInsertionEvent) throws Exception
     }
 
     if (((EnrichedEvent) tabletInsertionEvent).shouldParsePatternOrTime()) {
-      if (tabletInsertionEvent instanceof PipeInsertNodeTabletInsertionEvent) {
-        transfer(
-            ((PipeInsertNodeTabletInsertionEvent) tabletInsertionEvent).parseEventWithPattern());
-      } else { // tabletInsertionEvent instanceof PipeRawTabletInsertionEvent
-        transfer(((PipeRawTabletInsertionEvent) tabletInsertionEvent).parseEventWithPattern());
-      }
+      transfer((tabletInsertionEvent).parseEventWithPattern());
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/airgap/IoTDBAirGapConnector.java:190]: -

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+/**
+ * OPC UA Server builder for IoTDB to send data. The coding style referenced ExampleServer.java in
+ * Eclipse Milo.
+ */
+public class OpcUaServerUtils {
+
+  private static final String WILD_CARD_ADDRESS = "0.0.0.0";
+  private static int eventId = 0;
+
+  public static OpcUaServer getOpcUaServer(
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java:86]: Use "set" + "build"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+
+  /**
+   * Transfer tablet into eventNodes and post it on the eventBus, so that they will be heard at the
+   * subscribers. Notice that an eventNode is reused to reduce object creation costs.
+   *
+   * @param server OpcUaServer
+   * @param tablet the tablet to send
+   * @throws UaException if failed to create event
+   */
+  public static void transferTablet(OpcUaServer server, Tablet tablet) throws UaException {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java:276]: Move it into "connector"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+import static org.eclipse.milo.opcua.sdk.server.api.config.OpcUaServerConfig.USER_TOKEN_POLICY_ANONYMOUS;
+import static org.eclipse.milo.opcua.sdk.server.api.config.OpcUaServerConfig.USER_TOKEN_POLICY_USERNAME;
+import static org.eclipse.milo.opcua.sdk.server.api.config.OpcUaServerConfig.USER_TOKEN_POLICY_X509;
+import static org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.Unsigned.ubyte;
+
+/**
+ * OPC UA Server builder for IoTDB to send data. The coding style referenced ExampleServer.java in
+ * Eclipse Milo.
+ */
+public class OpcUaServerUtils {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java:81]: OpcUaServerBuilder

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+  private static final String WILD_CARD_ADDRESS = "0.0.0.0";
+  private static int eventId = 0;
+
+  public static OpcUaServer getOpcUaServer(
+      int tcpBindPort, int httpsBindPort, String user, String password) throws Exception {
+    final Logger logger = LoggerFactory.getLogger(OpcUaServerUtils.class);
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerUtils.java:88]: Move it into "Builder"

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/IoTDBKeyStoreLoader.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import java.security.KeyPair;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+import java.util.Set;
+import java.util.UUID;
+import java.util.regex.Pattern;
+
+class IoTDBKeyStoreLoader {
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/IoTDBKeyStoreLoader.java:43]: OpcUaKeyStoreLoader

**Review by Caideyipi — COMMENTED** [review]
Review

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerBuilder.java [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+ */
+public class OpcUaServerBuilder {
+
+  private final String wildCardAddress = "0.0.0.0";
+  private final Logger logger = LoggerFactory.getLogger(OpcUaServerBuilder.class);
+
+  private int tcpBindPort;
+  private int httpsBindPort;
+  private String user;
+  private String password;
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerBuilder.java:79]: A builder usually has default values for params.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerBuilder.java [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+  private static final String WILD_CARD_ADDRESS = "0.0.0.0";
+  private final Logger logger = LoggerFactory.getLogger(OpcUaServerBuilder.class);
+
+  private int tcpBindPort;
+  private int httpsBindPort;
+  private String user;
+  private String password;
+
+  public OpcUaServerBuilder() {
+    tcpBindPort = 12686;
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/connector/protocol/opcua/OpcUaServerBuilder.java:82]: org.apache.iotdb.db.pipe.config.constant.PipeConnectorConstant ?

**Pull request merged**